src/main/java/de/tum/in/www1/artemis/service/CourseService.java

```diff
@@ -674,4 +675,39 @@ public void checkIfEditorGroupsNeedsToBeCreated(Course course) {
             courseRepository.save(course);
         }
     }
+
+    /**
+     * Determines end date for the displayed time span of active student charts
+     * If the course end date is passed, only information until this date are collected and sent
+     * @param course the corresponding course the active students should be collected
+     * @return end date of the time span
+     */
+    public ZonedDateTime determineEndDateForActiveStudents(Course course) {
+        var endDate = ZonedDateTime.now();
+        if (course.getEndDate() != null && ZonedDateTime.now().isAfter(course.getEndDate())) {
+            endDate = course.getEndDate();
+        }
+        return endDate;
+    }
+
+    /**
+     * Determines the allowed time span for active student charts
+     * The span time can be restricted if the temporal distance between the course start date
+     * and the priorly determined end date is smaller than the intended time frame
+     * @param course the corresponding course the time frame should be computed
+     * @param endDate the priorly determined end date of the time span
+     * @param maximalSize the normal time span size
+     * @return the allowed time span size
+     */
+    public int determineTimeSpanSizeForActiveStudents(Course course, ZonedDateTime endDate, int maximalSize) {
+        var spanTime = maximalSize;
+        if (course.getStartDate() != null) {
+            var startDateIsoWeek = statisticsRepository.getWeekOfDate(course.getStartDate());
+            var endDateIsoWeek = statisticsRepository.getWeekOfDate(endDate);
+            int weeksInYear = Math.toIntExact(IsoFields.WEEK_OF_WEEK_BASED_YEAR.rangeRefinedBy(course.getStartDate()).getMaximum());
+            int amountOfWeeksBetween = (endDateIsoWeek - startDateIsoWeek + weeksInYear) % weeksInYear;
+            spanTime = Math.min(maximalSize, amountOfWeeksBetween + 1);
```

Comment:
I think there was something similar some time ago but I don’t remember the details: Why does
```java
course.getStartDate().until(endDate.plusWeeks(1), ChronoUnit.WEEKS)
```
not work in this case?

Reply:
I oriented at code that already performed a similar task. I tried your suggestion, which seemed to work for me, and passed the adapted test cases. Changed accordingly in b9ecc6d